### ISSUE TITLE
Deriving a store from props

### ISSUE THREAD

**Issue by Penspinner** [solkimicreb/react-easy-state#24]
My issue copied from Reddit.

What would be the solution to make Easy State work similar to `getDerivedStateFromProps`? What if I wanted a child component to update the store depending on the props passed from a parent component?

```javascript
class Parent extends Component {
  store = store({ number: Math.random() });

  handleButtonClick = () => {
      this.store.number = Math.random();
  };

  render() {
      return (
          <div>
              <button onClick={this.handleButtonClick}>Generate Random Number</button>                
              <Child number={this.store.number} />
          </div>
      );
  }
}

class Child extends Component {
  store = store({
      // number needs to update based on the number prop passed from parent on will receive props
      number: this.props.number + 1,
  });

  render() {
      return <div>{this.store.number}</div>;
  }
}
```

How to derive a store from props on will receive props?

----
solkimicreb
Ouch, I didn't think about that change :D I guess I could make it work exactly like it is now with two differences.

It would get the first local store on the component as the second argument, instead of the state.
It would merge the returned object with the first local store on the component, instead of the state.
Your example would look like this:

```javascript
class Child extends Component {
  store = store({
    number: this.props.number + 1
  });

  static getDerivedStateFromProps (props, store) {
    return { number: this.props.number + 1 };
  }

  render() {
    return <div>{this.store.number}</div>;
  }
}
```
This would place a few (tiny) restrictions on components with getDerivedStateFromProps.

These components should not have more than one local store to avoid confusion (of which store would receive the update object from getDerivedStateFromProps).
These components should not use vanilla React state in concjuction with local stores to avoid the above confusion.
Do you think this would be okay? Thanks for the catch (:

Edit: another approach would be to keep the exact current behavior with props and state and pass in a third argument, which is an array of local stores. Those stores could be mutated directly in the function, while the return value would be merged with state (as it is right now.) This feels a bit dirtier, but it would only extend the current behavior, instead of replacing some of it. Would this be better?

Edit2: I think I vouch for the first one, as the restrictions in places are also best practices. It might be beneficial to enforce them.

**Comment by solkimicreb** [comment]
Thanks for putting the issue here 🙂 I also read your other comment in Reddit since this.

I agree that it shouldn't be called `getDerivedStateFromProps`. I have two proposals:

**1. Don't do anything. Data should be derived with computed properties (getters).**

```js
class Child extends Component {
  constructor (props) {
    super(props)

    this.store = store({
      get number () {
        return props.number + 1
      }
    })
  }

  render() {
    return <div>{this.store.number}</div>
  }
}
```

I think this is a cleaner approach then having a separate static method, but it might be only my preference.

**2. Add a new `deriveStoresFromProps` static method. 

```js
class Child extends Component {
  store = store({})

  render() {
    return <div>{this.store.number}</div>
  }

  static deriveStoresFromProps(props, store) {
    store.number = props.number + 1
  }
}
```

`deriveStoresFromProps(props, ...stores)` would get the new props and all of the local stores of the component as arguments. It could then mutate the stores directly inside the function.

I am a bit afraid of the static nature of `getDerivedStateFromProps`. It might cause issues in the future, when it turns out why the React team decided to make it static. I suspect that they want to enforce pureness and idempotency with it.

Which one do you prefer? Both are easily doable, leaves vanilla React behavior untouched and includes now breaking changes.

**Comment by Penspinner** [comment]
The first approach is nice, but it's kind of a pain to make a getter for each property. Would the first approach reduce performance if the value is accessed many times continuously? Also, would it be possible to access the `this` if `store` was not in the constructor?
```
store = store({
  // Don't think the `this` refers to the component instance.
  number: this.props.number
});
```

For the second approach, would each mutation to the store trigger an update to the component? Would it be better for the second approach to be:
```
static getDerivedStoreFromProps(props, store) {
  if (props.number !== this.store.number) {
    // Store shallowly merged with current store.
    return store({ number: props.number });
  }

  return null;
}
```
Yes, the static nature makes it seem like they want to further enforce practices on state immutability.

**Comment by solkimicreb** [comment]
I wouldn't worry about performance for either approach, it won't be an issue. Renders are async batched, multiple synchronous mutations of the same store won't trigger multiple renders.

Yep, the first approach can only be used inside constructor functions. The second approach could work with a return value. I chose to do direct mutation, because this way we can support multiple local stores (there could only be a single return value). Honestly, I don't like it either way 😄  for now.

**Comment by Penspinner** [comment]
Dang, would like to create a store without having to create constructor.

Hm. I don't see why there would be multiple local stores in a component, similar to how each component should only have one state. I think it may be a good thing to enforce a rule of having a single local store in each component with `getDerivedStoreFromProps`.

I guess if you don't want to limit users to only having one local store, then `deriveStoresFromProps` would be fine.

**Comment by solkimicreb** [comment]
I have some prototypes for this, but I like neither of them 😄 I won't merge any of them before the stable React 16.3 release.

**Comment by solkimicreb** [comment]
I did not forget this issue 🙂 I am just waiting for things to settle. There are similar issues for other, bigger libraries - like mobx, but nobody came up with a good solution yet. I could do one of the workarounds, but it would totally 'unstaticize' the static method. I fear that it wouldn't go nicely with the React team's future plans.

**Comment by solkimicreb** [comment]
Hi @Penspinner!

I finally merged in the new `deriveStoresFromProps` lifecycle method and released v5.1.0. You can read more about it in [this docs section](https://github.com/solkimicreb/react-easy-state#how-do-i-derive-local-stores-from-props-getderivedstatefromprops).

Feel free to reopen the issue, if needed. 🙂
